File: src/main/java/net/jcflorezr/audiocontent/signal/RmsCalculator.java
package net.jcflorezr.audiocontent.signal;

import net.jcflorezr.model.audiocontent.signal.RmsSignalInfo;
import org.springframework.stereotype.Service;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

/**
 * The signal is divided into segments and for each segment the
 * <a href="http://en.wikipedia.org/wiki/Root_mean_square">RMS</a> value is computed.
 * For speech audio, a segment size of 100 ms might be reasonable.
 */
@Service
public class RmsCalculator {

    private static final double SILENCE_THRESHOLD = 0.001;
    private static final double ACTIVE_THRESHOLD = 0.03;

    public List<RmsSignalInfo> retrieveRmsInfo(float[][] signals, int segmentSize, int samplingRate) {
        int channel = 0;
        return retrieveRmsInfo(signals[channel], segmentSize, samplingRate);
    }

    private List<RmsSignalInfo> retrieveRmsInfo(float[] signal, int segmentSize, int samplingRate) {
        List<RmsSignalInfo> rmsSignalInfo = new ArrayList<>();
        int pos = 0;
        double prevRms = 0.0;
        double prevDiff = 0.0;
        while (pos < signal.length) {
            int endPos = (pos + segmentSize * 5 / 3 > signal.length) ? signal.length : pos + segmentSize;
            // If the last segment is less than 2/3 of the segment size, we include it in the previous segment.
            DecimalFormat df = new DecimalFormat("#.###");
            double rms = Double.parseDouble(df.format(computeRms(signal, pos, endPos - pos)));
            float positionInSeconds = (float) pos / samplingRate;

            double diff = Double.parseDouble(df.format(prevRms - rms));
            boolean silence = Math.abs(diff) <= SILENCE_THRESHOLD;
            double deepDiff = Double.parseDouble(df.format(prevDiff - diff));
            boolean active = Math.abs(deepDiff) >= ACTIVE_THRESHOLD;

            rmsSignalInfo.add(new RmsSignalInfo(rms, positionInSeconds, pos, silence, active));

            prevRms = rms;
            pos = endPos;
            prevDiff = diff;
        }
        return rmsSignalInfo;
    }

    private double computeRms(float[] signal, int startPos, int len) {
        double a = 0;
        for (int p = startPos; p < startPos + len; p++) {
            a += signal[p] * signal[p];
        }
        return Math.sqrt(a / len);
    }

}

File: src/main/kotlin/net/jcflorezr/rms/RmsCalculator.kt
package net.jcflorezr.rms

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import mu.KotlinLogging
import net.jcflorezr.broker.Topic
import net.jcflorezr.model.AudioSignal
import net.jcflorezr.model.AudioSignalRmsInfo
import net.jcflorezr.model.AudioSignalsRmsInfo
import net.jcflorezr.util.AudioUtils
import net.jcflorezr.util.PropsUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

interface RmsCalculator {
    suspend fun generateRmsInfo(audioSignal: AudioSignal)
}

@Service
class RmsCalculatorImpl : RmsCalculator {

    @Autowired
    private lateinit var audioSignalRmsTopic: Topic<AudioSignalsRmsInfo>

    private val logger = KotlinLogging.logger { }

    companion object {
        private const val SILENCE_THRESHOLD = 0.001
        private const val ACTIVE_THRESHOLD = 0.03
    }

    override suspend fun generateRmsInfo(audioSignal: AudioSignal) {
        logger.info { "[${PropsUtils.getTransactionId(audioSignal.audioFileName)}][3][RMS] " +
            "Start to generate the Root Mean Square (RMS) of audio signal start: ${audioSignal.initialPositionInSeconds} - " +
            "end: ${audioSignal.endPositionInSeconds}." }

        val sampleRate = audioSignal.sampleRate
        val segmentSize = sampleRate / 10 // 0.1 secs
        val signal = audioSignal.data[0]!! // always mono
        val rmsSignalsInfo = ArrayList<AudioSignalRmsInfo>()
        generateSequence(RmsValues(pos = 0, index = 1, rms = 0.0, diff = 0.0)) {
            // If the last segment is less than 2/3 of the segment size, we include it in the previous segment.
            val endPos = if (it.pos + segmentSize * 5 / 3 > signal.size) signal.size else it.pos + segmentSize
            val currentRms = AudioUtils.millisecondsFormat(
                    value = computeRms(signal = signal, startPosition = it.pos, length = endPos - it.pos)
                )
            val positionInSeconds = it.pos.toFloat() / sampleRate
            val currentDiff = AudioUtils.millisecondsFormat(value = (it.rms - currentRms))
            val deepDiff = AudioUtils.millisecondsFormat(value = (it.diff - currentDiff))
            val silence = Math.abs(currentDiff) <= SILENCE_THRESHOLD
            val active = Math.abs(deepDiff) >= ACTIVE_THRESHOLD
            val initialPosInSeconds = positionInSeconds + audioSignal.initialPositionInSeconds
            val rmsSignalInfo = AudioSignalRmsInfo(
                audioFileName = audioSignal.audioFileName,
                index = AudioUtils.tenthsSecondsFormat(initialPosInSeconds.toDouble()),
                rms = currentRms,
                audioLength = audioSignal.totalFrames,
                sampleRate = sampleRate,
                initialPosition = it.pos + audioSignal.initialPosition,
                initialPositionInSeconds = initialPosInSeconds,
                segmentSize = segmentSize,
                segmentSizeInSeconds = AudioUtils.tenthsSecondsFormat(segmentSize / sampleRate.toDouble()).toFloat(),
                silence = silence,
                active = active
            )
            rmsSignalsInfo.add(rmsSignalInfo)
            RmsValues(pos = endPos, index = it.index + 1, rms = currentRms, diff = currentDiff)
        }.takeWhile { it.pos < signal.size }
        .toList()
        logger.info { "[${PropsUtils.getTransactionId(audioSignal.audioFileName)}][3][RMS] " +
            "Root Mean Square (RMS) for audio signals ==> start: ${rmsSignalsInfo.first().initialPositionInSeconds} - " +
            "end: ${rmsSignalsInfo.last().initialPositionInSeconds} were generated." }
        audioSignalRmsTopic.postMessage(message = AudioSignalsRmsInfo(rmsSignalsInfo))
    }

    private fun computeRms(signal: FloatArray, startPosition: Int, length: Int): Double {
        var a = 0.0
        for (p in startPosition until startPosition + length) {
            a += (signal[p] * signal[p]).toDouble()
        }
        return Math.sqrt(a / length)
    }

}

private data class RmsValues(
    val pos: Int,
    val index: Int,
    val rms: Double,
    val diff: Double
)
